### PR TITLE
Sam refactor

### PR DESCRIPTION
**Pull request by sminnee**
I'm refactoring things to provide an abstract base class for the different animations / patterns. This makes it easier to swap around the logic for how patterns are used more easily.

@bnolan @stojg - keen to get some feedback on whether this makes sense

### CODE REVIEW

**Comment by sminnee** [comment]
OK I'm pretty happy with this for now – would be good to have a read over it and merge thanks :-)

File: control-box/ASRFader.h [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ * A fader with fade-in, sustain, and fade-out. Extended from the Light Hoops' PIRSensor.
+ * 
+ * This works well with a PIR motion sensor. Set it for minimum cycle time, so they have
+ * a period of about 1-2 seconds. Use this to sustain patterns for the appropriate duration.
+ *
+ * With the fade in & out, you can build smooth transitions between the "on" and "off" states.
+ * 
+ * @author Sam Minnee
+ */
+class ASRFader {
```
Comment (stojg) [inline control-box/ASRFader.h:17]: So this is the state machine that keeps track of the PIR status and keeps track of the fade value?

File: control-box/PatternList.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
+#include "Pattern.h"
+#include <FastLED.h>
```
Comment (stojg) [inline control-box/PatternList.h:2]: does this need a
```
#ifdef FASTLED
#include <FastLED.h>
#define FASTLED
#endif
```

Macro? My c++ linking skills is a bit rusty.

Comment (stojg) [reply control-box/PatternList.h:2]: Or would it be better to put the include in Pattern.h ?

File: control-box/ASRFader.h (hunk 1 above)
Comment (sminnee) [reply control-box/ASRFader.h:17]: yes, that's right.

File: control-box/PatternList.h (hunk 2 above)
Comment (sminnee) [reply control-box/PatternList.h:2]: I assume that FastLED already has an ifdef block of that kind within it.

Yes, I could just list it in PatternState.h (the root dependency)

**Comment by bnolan** [comment]
Stig reviewed, looks good to him.

**Pull request merged**